builder/comp-builder.nix

```diff
@@ -166,14 +163,59 @@ let
   exeName = componentId.cname + exeExt;
   testExecutable = "dist/build/${componentId.cname}/${exeName}";
 
-in stdenv.lib.fix (drv:
+  # Attributes that are common to both the build and haddock derivations
+  commonAttrs = {
+      src = cleanSrc;
+
+      LANG = "en_US.UTF-8";         # GHC needs the locale configured during the Haddock phase.
+      LC_ALL = "en_US.UTF-8";
+
+      enableParallelBuilding = true;
+
+      SETUP_HS = setup + /bin/Setup;
+
+      prePatch = if (cabalFile != null)
+         then ''cat ${cabalFile} > ${package.identifier.name}.cabal''
+         else lib.optionalString (cabal-generator == "hpack") ''
+           ${buildPackages.haskell-nix.haskellPackages.hpack.components.exes.hpack}/bin/hpack
+         '';
+    }
+    # patches can (if they like) depend on the version and revision of the package.
+    // lib.optionalAttrs (patches != []) {
+      patches = map (p:
+        if builtins.isFunction p
+          then p { inherit (package.identifier) version; inherit revision; }
+          else p
+        ) patches;
+    }
+    // haskellLib.optionalHooks {
+      inherit preUnpack postUnpack preConfigure postConfigure
+        preBuild postBuild preHaddock postHaddock
+        preInstall postInstall;
+    }
+    // lib.optionalAttrs (stdenv.buildPlatform.libc == "glibc") {
+      LOCALE_ARCHIVE = "${buildPackages.glibcLocales}/lib/locale/locale-archive";
+    };
+
+in
+
+stdenv.lib.fix (drv:
```

Comment:
Perhaps stupid question, but why not just use a recursive let-binding here?

Reply:
I sort of like that you only have to specify drv once.

```
in stdenv.lib.fix (drv: ... )
```

vs

```
in let rec drv = ...; in drv
```

Also it might be too easy to add other stuff to the `let rec` and then forget it is not a regular `let`.

I have no strong feeling on this though and I'm happy to change it.

Reply:
Oh I am a twit. All `let`s are `rec` in nix.  I'll change it.

Reply:
I don't particularly disagree that it's nice to have more control over the recursivity, it's just that doing it this way is a little unusual, so I'd expect a comment at least.

Reply:
I'm going to switch it in a separate PR as it requires indenting a lot of code that will likely make it harder to find the changes of this PR.